certbot/log.py

```diff
@@ -240,9 +238,12 @@ class TempHandler(logging.StreamHandler):
 
     """
     def __init__(self):
-        stream = tempfile.NamedTemporaryFile('w', delete=False)
+        fd, path = tempfile.mkstemp()  # To get a proper tempfile path
```

Comment:
Could you explain what was wrong with the original code here?

Reply:
Wrong permissions on Windows, since `NamedTemporaryFile` is calling internally `os.open`, that is creating wrong DACL and owner for a file. Since this class explicitly requires secured permissions through its unit tests, I changed its implementation to use the `filesystem` methods.

Reply:
Hold up, how was `NamedTemporaryFile` able to call `os.open` internally without throwing an error? Is our "replace `os.open` in `sys.modules`" plan not working properly?

Reply:
But, it is not the plan. We discussed that and decided together not monkey patching os module.

Reply:
Still not clear to me, I do not understand what you are asking me here.

I replaced here a call to a function from builtin module `tempfile` to another function from this same builtin module. Any call from functions of `tempfile` cannot it our `certbot.compat.os` function, since `tempfile` is not from our codebase.

I could use `filesystem.open` directly here, but I would loose the capability of `tempfile.mkstemp` to use the OS APIs to create consistently a new temporary file with random name, in particular on Linux. Instead, I first call `tempfile.mkstemp` to get the path, then delete it, then re-create it correctly using `util.safe_open` both for Windows and Linux.

If you are not satisfied with this approach, I propose to create first a temporary _directory_ using `tempfile.mkdtemp()`, and not care about the permissions on this directory, then create a safe `log` file in it using `util.safe_open`, and use this file as the stream facility for the `TempHandler`.

Reply:
I implemented the logic described in the last paragraph in by comment above.

Reply:
To ask in a different way: isn't the point of replacing `os.open` in `sys.modules` with `certbot.compat.os.open` that calls to `os.open` error out? Is there a different `sys.modules` for different namespaces? Why isn't code in `tempfile` using our modified `sys.modules`?

Reply:
No it is not the point. We decided to not try to make a perfect replacement of `os.open` or any other function from `os` module, and not replace the `os` module to `certbot.compat.os` in the python runtime of Certbot.

Instead we build functions suitable and crafted explicitly for Certbot use, and replace only the call to `os` done in our own codebase only. We do not alterate in any way the behavior of our third party dependencies, or the Python distribution itself. If this leads to undesired results for Certbot, we replace these faulty calls by calls to functions from our own codebase. This is the case here with `tempfile` for this implementation of `TempHandler` that requires a strictly secure access.

Reply:
In that case, could you explain what these lines do? https://github.com/certbot/certbot/blob/master/certbot/compat/os.py#L21-L27

Reply:
They are copying every member of the `os` module into the `certbot.compat.os` module (including private members and everything not in the special `__all__` dict of the module), in order to use the `certbot.compat.os` as a fully functional `os`-like module in our codebase.